### ISSUE TITLE
Middleware: implement "next" callback

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#4039]
Unlike Starcounter, most middleware-oriented frameworks, like Koa.js or ASP.NET Core, don't have a concept of "request" and "response" middlewares. 

There is only one type of middleware with `next` callback function.

It is explained well in ASP.NET Core [docs](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware):

> Each delegate has the opportunity to perform operations before and after the next delegate. Any delegate can choose to stop passing the request on to the next delegate, and instead handle the request itself. This is referred to as short-circuiting the request pipeline, and is desirable because it allows unnecessary work to be avoided. For example, an authorization middleware might only call the next delegate if the request is authenticated; otherwise it could short-circuit the pipeline and return a "Not Authorized" response. Exception handling delegates need to be called early on in the pipeline, so they are able to catch exceptions that occur in deeper calls within the pipeline.

**Comment by per-samuelsson** [comment]
@warpech,

When we brought the middleware concept to Starcounter, we did so initially to support it as the API over already existing "non-standard" request- and response filters. I'm sure you remember this. At that time, we decided to postpone next callback until a later version, possible 3.x. Basically because the underlying bits worked as they did (and it was kind of messy to change/redesign that).

We could certainly consider doing this now, and have it already in 2.4+ < 3.x, but let's discuss first if that's what we want, if it's not better to just wait until we get to 3.x (where we build on ASP.NET Core and their middleware pipeline).

**Comment by per-samuelsson** [comment]
@warpech 

I had a vague feeling I actually started adopting some kind of delegate chaining (i.e `next()`) already the last time around, but I searched and couldn't find any issue on it. But now I found my old "article" and particularly [this section](http://starcounter.io/guides/network/request-response-filters/#json-view-models-html-views), and there it looks like we might already support it to some level. 😕 😎 Maybe I did something after all?

**Comment by warpech** [comment]
Yes, you did it actually in [MimeProvider](https://github.com/Starcounter/Starcounter/blob/81c9a245f2f3769cb19bb2e2ad0c3d69a14b19b8/src/Starcounter/Middleware/MimeProvider.cs).

I think it should be moved to `Application.Use` if possible at all.
